HDDS-6449. Failed container delete can leave artifacts on disk

xBis7 opened the pull request:
## What changes were proposed in this pull request?

In case there is a failure during a container delete there will be artifact leftovers on disk. To avoid that, containers will be renamed to a new location under a tmp directory and will be deleted from that location. This solution only works if SchemaV3 is enabled. For previous Schema versions, delete operation is exactly the same.

To facilitate this implementation, a helper class called `CleanUpManager` was created which initializes the tmp directory on each disk under the path `/<volume>/<clusterId>/tmp/container_delete_service` and handles all operations related to this directory. Before every container delete, if SchemaV3 is enabled then we rename the container, set the path to the new location and then delete it from there. 

On datanode startup and shutdown we are checking the tmp directory for any leftovers and then clean it up. The datanode startup hook is in `VersionEndpointTask.checkVolumeSet()` and the shutdown is in `HddsDatanodeService.stop()`.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6449

## How was this patch tested?

This patch was tested with new unit tests. We added `TestContainerPersistence.testDeleteContainerWithSchemaV3Enabled()` for testing `CleanUpManager` and `TestKeyValueHandler.testDeleteContainerWithSchemaV3Enabled()` for testing `KeyValueHandler.deleteInternal()`. In addition, we tested cleaning the tmp directory on datanode startup in `TestEndPoint.testGetVersionTask()` and on datanode shutdown in `TestHddsDatanodeService`. 

There were some issues with the existing code for `TestContainerPersistence.testDeleteContainer()`. The code block that was expected to throw an non-empty container exception, was never executed while the method was exiting upon the first exception thrown. Furthermore, the command that was used for deleting the container wasn’t checking if it was empty or not. In order to fix that, the method was splitted into two new ones and `KeyValueHandler.deleteContainer()` was used instead of `container.delete()`. 

Due to the changes related to SchemaV3 some tests under `container-service` package were failing, so we had to make small changes to fix their setups.

xBis7 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/CleanUpManager.java:
```diff
@@ -0,0 +1,207 @@
+      clusterId += hddsVolume.getClusterID();
+
+      if (clusterId == null) {
+        throw new IOException();
```
@myskov In the catch statement we are logging both the message and the exception.

myskov replied:
@xBis7 yes, but it's better for error description to be part of the exception. If catch statement will be changed in the future, this exception may be left as is.

myskov commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/helpers/CleanUpManager.java:
```diff
@@ -0,0 +1,214 @@
+  private static final String TMP_DELETE_SERVICE_DIR =
+      FILE_SEPARATOR + "tmp" + FILE_SEPARATOR + "container_delete_service";
+
+  private Path tmpDirPath;
```
can be final?

errose28 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java:
```diff
@@ -556,6 +569,9 @@ public void terminateDatanode() {
 
   @Override
   public void stop() {
+    if (CleanUpManager.checkContainerSchemaV3Enabled(conf)) {
+      cleanTmpDir();
+    }
```
The shutdown hook should be moved inside the atomic boolean check to make sure it is only invoked once.

errose28 reviewed (commented):
Thanks for working on this @xBis7. I have not reviewed the tests yet. I left some more specific comments in line but I have some high level thoughts on this as well:

- I do not see anything specific to schema V3 in the deletion changes. Why are we only allowing it for schema V3?
- EC offline recovery will need its own tmp subdirectory with its own handling rules. I am not sure how it works in master right now but we can leave any code related to RECOVERING containers as is for this PR and handle EC tmp directory usage as a separate task.
- Since we are adding a new directory on datanode startup, can you run the full upgrade/downgrade test suite to check for compatibility? Details on how to do that are [here](https://github.com/apache/ozone/pull/3648#issuecomment-1211047569).
- I think current changes could be improved with some refactoring keeping the following points in mind:
    - The tmp directory belongs to the volume. It should be created when the volume is initialized.
    - The tmp directory will need to be used in a lot of places in the future: container create, copy, delete, and EC offline recovery. It should be easy for any of these services to get the tmp path if they have a volume to operate on, and to add their own operating directory under it as needed. This probably means adding it as a property of `HddsVolume`, or letting it be built with a method in `HddsVolumeUtil` depending on whether you want to require an actual instance of the volume to get the directory. 
    - I do not think the `CleanUpManager` class is necessary.
        - Its only state is the path to the tmp delete directory, which could be obtained with a new method in either `HddsVolumeUtil` or `HddsVolume` as stated above.
        - Once the EC changes are removed, methods that are called only once can be moved to the class that is calling them.
        - Methods that are called only for testing can be moved to the test requiring them, or to a generic testing util class if needed in multiple tests.
        - After these changes I think `cleanTmpDir` is the only method needed in two places: the shutdown and startup hooks. In both of those places we have `HddsVolume` objects so it can be moved in to that class. `HddsVolumeUtil` could be used as well.

xBis7 commented:
@errose28 Thanks for reviewing this. I will make the changes and modify the tests accordingly. Prior to `SchemaV3`, the clusterId doesn't appear to be available. We were told that clusterId should be available in the tmp dir path and that was an issue with previous versions.

errose28 commented:
The cluster ID directory on the datanode has a complicated history, but it is not related to schema V3. For a detailed look see HDDS-5432 and the design document linked there. Basically before SCM HA the cluster ID directory on each volume was actually using the SCM ID. This became a problem when SCM HA was introduced since there are now 3 SCMs, but renaming the directory was an incompatible change. At the end of it all the volume could be formatted in the following ways:
- Only cluster ID directory
    - The volume was formatted with Ozone 1.3.0
- SCM ID directory pointed to by a cluster ID symlink
    - The volume was formatted with Ozone <= 1.2.0 and SCM HA was finalized.
- Only SCM ID
    - The volume was formatted with Ozone <= 1.2.0 and SCM HA has not yet been finalized.
    - This would happen on the upgrade path from 1.1.0 to 1.3.0

In general you should not need to worry about this. Just call `VersionedDatanodeFeatures#ScmHA#chooseContainerPathID` and use the result as the parent directory.

errose28 commented:
One other thing to check is what happens with schema V3 when the container is moved to the tmp dir but the RocksDB update fails. Will the datanode restart cleanly? What happens to the RocksDB entries for that container?

xBis7 commented:
@errose28 Thanks for your comments, I've made the changes and also modified the tests to use the new code path. Let me know what you think.

> One other thing to check is what happens with schema V3 when the container is moved to the tmp dir but the RocksDB update fails. Will the datanode restart cleanly? What happens to the RocksDB entries for that container?

I will be looking at that next.

neils-dev commented:
Thanks @errose28 for our offline discussions on this container delete / rename change and for the PR review and comments.

> One other thing to check is what happens with schema V3 when the container is moved to the tmp dir but the RocksDB update fails. 

During the design discussions with @swagle and @nandakumar131 on this case for failed access to db to delete container on container delete operation from, 
https://github.com/apache/ozone/blob/0e9ea9fba0d78becc393945205fa5450ab31f066/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java#L149
`BlockUtils.removeContainerFromDB(containerData, conf);` it was deemed to be unrecoverable.
At the time we called it an unrecoverable error and should the access to delete the container from the db fail, then the disk should be replaced.

Right now for that case, we log the error and throw an exception that is translated to a `StorageContainerException`.  Presently, there is no recoverable path for the db container delete failure like was discussed in the design discussions.  If it should be recoverable, what do you suggest?   @swagle , @nandakumar131 thoughts?

errose28 commented:
I've looked through the existing code, and if the RocksDB update fails for schema v3 I think we will be ok with a minor improvement in the current design. The volume may eventually be marked unhealthy but even if it is not I think this error is recoverable.

## Schema < V3

For schema < V3 we have the following operations:
1. Container directory renamed to tmp directory.
2. Container is removed from in memory container set.
3. Container is deleted from tmp directory.

We had initially designed to handle a failure in steps 1 or 3. Step 2 is in memory so it should fail. For schema V3, however, we have an extra operation since the volume's RocksDB needs to be updated.

## Schema V3

For schema V3 we have the following operations:
1. Container directory renamed to tmp directory.
2. Container is removed from in memory container set.
3. Container's entries are removed from RocksDB.
4. Container is deleted from tmp directory.

The original design still handles failures in steps 1 and 4 above as intended, but what happens if step 3 fails? The container has been removed from the in memory container set in step 2, so there should be any correctness issues while the datanode is running. On restart, the container set is rebuilt by scanning the available datanode volumes, and container metadata is populated based on the values in RocksDB. Since the container is already in the tmp directory, it will not be seen in this step and will not be loaded in to the container set, so we are still behaving correctly.

This does leave some orphaned entries in the volume's RocksDB, however I think these should be relatively easy to clean up. On the startup/shutdown hooks where we are deleting the containers from the tmp directory, we can check if the container is schema V3. If it is, we should first delete any remaining RocksDB entries it may have (if they exist) and then we can delete the container from the tmp directory. If this RocksDB delete fails the container should be skipped and its delete will happen again on the next startup/shutdown. Since the container move to the tmp directory is atomic, we can check RocksDB for each schema V3 .container file we find in the tmp directory. If the contents of the tmp directory are partially deleted and there is no .container file  (maybe a failed delete from the tmp directory in the past) then the RocksDB entries must have already been removed so we can proceed to delete those contents.

errose28 reviewed (commented):
Thanks for the updates @xBis7. I think my original comment on refactoring was not clear. I meant that the <volume>/hdds/<clusterID>/tmp directory itself should be maintained in a central place, like HddsVolume or HddsVolumeUtil, which the updated code has. However operations on the deletion subdirectory should be managed where the deletion actually happens, like KeyValueHandler or KeyValueContainerUtil. This way the tmp directory can be used by all operations that need it in the future like create and import, but the HddsVolume classes remain as general storage locations not concerned with container operations.

kerneltime commented:
cc @aswinshakil

captainzmc commented:
Hi @errose28 , In dev email communication, this PR is required by Ozone 1.3. Currently the release of Ozone-1.3 is blocked on this PR. Could you help continue to review this PR?

errose28 commented:
Yes, I'm working on a re-review to get this through.

errose28 reviewed (commented):
Thanks for the updates @xBis7 and sorry I took so long to come back to this. Did you get a chance to try the downgrade to 1.1.0 testing I mentioned earlier?

errose28 commented:
@captainzmc we still need some more work to get this PR ready. I'm not sure it needs to be a release blocker. The bug is if there is an IO error while deleting the container, some pieces may be left behind in the container directory. This will cause the datanode to log a warning on startup but has no impact on functionality. Users concerned about the log message can check the container ID using SCM admin commands to see that it is deleted and manually remove the artifacts safely if they wish.

xBis7 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/states/endpoint/VersionEndpointTask.java:
```diff
@@ -122,6 +124,19 @@ private void checkVolumeSet(MutableVolumeSet volumeSet,
             ozoneContainer.getDbVolumeSet());
+
+        // Clean <HddsVolume>/tmp/container_delete_service dir.
+        if (volume instanceof HddsVolume) {
```
> I would say to not fail the volume, as the tmp dir being clean on startup is not required for functionality.

Tmp dir will also be cleaned on shutdown. I agree, let's leave it as it and just `LOG.error()` upon catching the exception.

xBis7 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:
```diff
@@ -416,6 +418,8 @@ ContainerCommandResponseProto handleDeleteContainer(
       deleteInternal(kvContainer, forceDelete);
     } catch (StorageContainerException ex) {
       return ContainerUtils.logAndReturnError(LOG, ex, request);
+    } catch (IOException ex) {
```
This IOException comes from [here](https://github.com/apache/ozone/blob/78447eb08463a89a2ad0c75d5ce0b2e3597705a6/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java#L568). I don't think we should surpress this or change it to `StorageContainerException`.

xBis7 replied:
@errose28 Just logged the error and removed the exception alltogether, since you mentioned that container and volume scans will deal with it.

errose28 replied:
We still need to throw the StorageContainerException if the move fails so that SCM will retry the delete. My point about the scanners was that it is ok if SCM keeps retrying even though there is some IO issue with the volume, because the scanners will eventually pick up on it and fail the volume, after which the deletions will not be sent again.

xBis7 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:
```diff
@@ -395,4 +416,174 @@ public static Path getMetadataDirectory(
+     * @return container file or null if it doesn't exist
+     * @throws IOException
+     */
+    public static File getContainerFile(File file) throws IOException {
```
@errose28 There is a chance that cleaning the tmp directory can fail and then the container directory won't be in it's normal structure and format. In that case, ContainerUtils.getContainerFile() won't work as expected. That's why I had created my own method that searches recursively the directory for the container file.

xBis7 replied:
> What would be an example of this failure case?

I don't really remember what I was thinking at that point.

> The delete does not move this file relative to the base location.

You are right. I'll make the change and fix the tests.

xBis7 commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestContainerPersistence.java:
```diff
@@ -286,19 +288,161 @@ public void testDeleteContainer() throws Exception {
+    interval[0] = 2;
+    ContainerMetrics metrics = new ContainerMetrics(interval);
+
+    AtomicInteger icrReceived = new AtomicInteger(0);
```
A reasonable test would be adding `Assert.assertEquals(0, icrReceived.get());` after `kvHandler.deleteContainer()` to check that the operation failed, but it's redundant. After line 328 runs, the exception gets thrown and nothing executes below that.

errose28 replied:
I don't see the value getting read anywhere so I was just wondering if it was being used in the lambda somehow and I missed it.

xBis7 replied:
No, we just need it to initialize `KeyValueHandler`.

captainzmc commented:
> @captainzmc we still need some more work to get this PR ready. I'm not sure it needs to be a release blocker. The bug is if there is an IO error while deleting the container, some pieces may be left behind in the container directory. This will cause the datanode to log a warning on startup but has no impact on functionality. Users concerned about the log message can check the container ID using SCM admin commands to see that it is deleted and manually remove the artifacts safely if they wish.

Thanks to @errose28  for explaining the impact of this bug. It seems that this does not affect the use of the cluster, and the user can manually delete these artifacts.  
Hi @neils-dev.  The release of 1.3.0 is now nearly two months behind schedule. If this PR takes too long to improve, I recommend removing it from the block list. In addition, Some new bugs may be found after the 1.3.0 release. So we can cherry pick the new bug fix from the master later, and then we can continue to release 1.3.1. What do you think?

xBis7 commented:
> If this PR takes too long to improve, I recommend removing it from the block list. In addition, Some new bugs may be found after the 1.3.0 release. So we can cherry pick the new bug fix from the master later, and then we can continue to release 1.3.1

@captainzmc okay, let's do that.

neils-dev commented:
Thanks @captainzmc, @errose28 , @xBis7 ,

> Hi @neils-dev. The release of 1.3.0 is now nearly two months behind schedule...  I recommend removing it from the block list

Sure, let's remove it from the block list.  Thanks!

xBis7 commented:
@errose28 I'm done with the changes you requested. I left a few comments based on some details that are unclear to me.

xBis7 commented:
> Did you get a chance to try the downgrade to 1.1.0 testing I mentioned earlier?

@errose28 I was able to run this script successfully twice `hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/test.sh`, once with `run_test non-rolling-upgrade 1.1.0 1.3.0` and once with `run_test non-rolling-upgrade 1.2.1 1.3.0`. For some weird reason, I had to install robotframework under `/compose/upgrade` to get it to work, even though I have it installed in my machine. Apart from that, no issues. I also checked the result logs and I couldn't find any errors related to our changes.

xBis7 commented:
> Did you get a chance to try the downgrade to 1.1.0 testing I mentioned earlier?

@errose28 I was able to run this script successfully twice `hadoop-ozone/dist/target/ozone-1.3.0-SNAPSHOT/compose/upgrade/test.sh`, once with `run_test non-rolling-upgrade 1.1.0 1.3.0` and once with `run_test non-rolling-upgrade 1.2.1 1.3.0`. For some weird reason, I had to install robotframework under `/compose/upgrade` to get it to work, even though I have it installed in my machine. Apart from that, no issues. I also checked the result logs and I couldn't find any errors related to our changes.

errose28 reviewed (commented):
Hi @xBis7 sorry again for delays around this PR. A few more comments and we should double check on [this conversation](https://github.com/apache/ozone/pull/3741#discussion_r1036768130) as well.

xBis7 commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:
```diff
@@ -88,6 +130,67 @@ public void testStartup() throws IOException {
+  @ParameterizedTest
+  @ValueSource(strings = {OzoneConsts.SCHEMA_V1,
+      OzoneConsts.SCHEMA_V2, OzoneConsts.SCHEMA_V3})
+  public void testTmpDirOnStartup(String schemaVersion) throws IOException {
```
You are right. I got carried away by the name of the other method. I'll fix it!

xBis7 commented:
@errose28 thanks for reviewing this. I made the changes and updated the tests. Also, added an scm exception in case moving the container failed. Let me know how it looks.

neils-dev reviewed (commented):
Thanks @xBis7  for this patch to fix the container delete disk artifacts problem.  Thanks @errose28 for your help and review comments to finalize this design and implementation. 

From going over the conversations in this PR, it appears that the main comments have been discussed and addressed, those being,
i.) refactoring the container delete  tmp directory path related items into the `hddsVolume` and placing the container move related items into the` keyValueContainerUtils`.

ii.) expanding the container delete move to tmp to include datanode schema support pre Schema V3. (support for legacy pre RocksDB per volume)

iii.) handling error case where for V3 schema, the RocksDB entry for the container is not deleted on move.  On `start/shutdown handle`r container delete from tmp directory, check and _delete outstanding RocksDB entries_ for containers found in tmp directory.

iv.) testing the affect of container delete by atomic move to tmp directory on older versions of ozone.  Using the downgrade to 1.1.0 non-rolling upgrade docker cluster to test.  @xBis7 reported no issues with the downgrade.

v.) reusing the `ContainerUtils` instead of adding new method to g`etContainerFile `

vi.) throwing a `StorageContainerException` in the `KeyValueHandler#deleteInternal `on container move to tmp directory failure.

@errose28 please have a look at the updates to the patch.  It looks like all comments have been addressed.  @xBis7 at my newly added comments.  

If no further comments, I will look to approve and merge the changes in the next few days.  Thanks!

xBis7 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:
```diff
@@ -1239,6 +1240,29 @@ private void deleteInternal(Container container, boolean force)
+          LOG.error("Failed to move container under " +
+              hddsVolume.getDeleteServiceDirPath());
+          throw new StorageContainerException("Moving container failed",
+              IO_EXCEPTION);
```
@neils-dev Thanks for the suggestion. I made the change.

errose28 commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:
```diff
@@ -88,6 +131,75 @@ public void testStartup() throws IOException {
+        volumeSet.getVolumesList());
+
+    for (HddsVolume volume : volumes) {
+      volume.format(clusterId);
+      volume.createWorkingDir(clusterId, null);
+      volume.createDeleteServiceDir(clusterId);
```
It might be better to call `StorageVolumeUtil#checkVolume` here instead to make the test resemble volume init more accurately.

errose28 replied:
Similar comment for TestEndPoint#testTmpDirCleanup

errose28 commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:
```diff
@@ -88,6 +131,75 @@ public void testStartup() throws IOException {
+    service.join();
+    service.close();
+
+    for (HddsVolume hddsVolume : volumes) {
+      ListIterator<File> deleteLeftoverIt = KeyValueContainerUtil
+          .ContainerDeleteDirectory.getDeleteLeftovers(hddsVolume);
+      assertFalse(deleteLeftoverIt.hasNext());
+    }
```
It would be good to have a similar block before the shutdown to make sure that the test was set up properly and the container made it to the tmp directory.

errose28 replied:
Similar comment for TestEndPoint#testTmpDirCleanup

errose28 commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/ozone/container/common/TestEndPoint.java:
```diff
@@ -165,17 +165,95 @@ public void testGetVersionTask() throws Exception {
         true);
     conf.setBoolean(OzoneConfigKeys.DFS_CONTAINER_RATIS_IPC_RANDOM_PORT,
         true);
-    conf.setBoolean(OzoneConfigKeys.DFS_CONTAINER_RATIS_DATASTREAM_ENABLED,
```
Looks like this test had ratis streaming enabled but it got turned off by accident. May have happened when resolving conflicts from master.

errose28 replied:
I see a similar thing in `testGetVersionTask`

xBis7 replied:
Yes, that must have been an accident. I only meant to cleanup the `setUp()`.

errose28 commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/ozone/container/common/TestEndPoint.java:
```diff
@@ -165,17 +165,95 @@ public void testGetVersionTask() throws Exception {
+    }
+  }
+
+  private KeyValueContainer setUpTestContainer(
```
This looks  similar to `TestHDDSDatanodeService#setUpTestContainer`. Can we make these one method and move it to a common place like `ContainerTestUtils`?

errose28 commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/ozone/container/common/TestEndPoint.java:
```diff
@@ -105,20 +116,13 @@ public static void setUp() throws Exception {
     scmServer = SCMTestUtils.startScmRpcServer(SCMTestUtils.getConf(),
         scmServerImpl, serverAddress, 10);
     testDir = PathUtils.getTestDir(TestEndPoint.class);
-    config = SCMTestUtils.getConf();
-    config.set(DFS_DATANODE_DATA_DIR_KEY, testDir.getAbsolutePath());
-    config.set(OZONE_METADATA_DIRS, testDir.getAbsolutePath());
-    config
-        .setBoolean(OzoneConfigKeys.DFS_CONTAINER_RATIS_IPC_RANDOM_PORT, true);
-    config.set(HddsConfigKeys.HDDS_COMMAND_STATUS_REPORT_INTERVAL, "1s");
-    config.setFromObject(new ReplicationConfig().setPort(0));
```
Why were these configs removed?

xBis7 replied:
@errose28 These configs were unecessary for the existing tests, so I thought about cleaning them up. If there is something I missed, I will put them back.

errose28 reviewed (commented):
I have some more comments but they are smaller. I think this is nearing completion.

xBis7 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/HddsVolume.java:
```diff
@@ -326,6 +345,70 @@ public void createDbStore(MutableVolumeSet dbVolumeSet) throws IOException {
+   * 
+   * @param id clusterId or scmId
+   */
+  public void checkTmpDirPaths(String id) {
```
@errose28 Tmp directories are only created when needed and not along with volume initialization. Without this method setting the paths everything breaks.

xBis7 replied:
@errose28 The tmp is created when creating a working dir. There are cases were this step is missed because working dir already exists but the paths used for the tmp dir are null and we need to set them again. Check a workflow I launched where we are skipping this method.

https://github.com/xBis7/ozone/actions/runs/3874176967/jobs/6605289826

neils-dev replied:
@xBis7 , we can remove the need for the `checkTmpDir` methods in the `HddsVolume` with a few changes that cleanup the code and also resolves this issue.  
1.) remove `HddsVolume.checkTmpPaths` 
2.) remove `checkCleanupDirs `method and calls in `StorageVolumeUtil`
3.) in `HddsVolume.createTmpPath` get the `id` from a call to new` HddsVolume` method,
`private String getIdDir() `that returns versions id : `return VersionedDatanodeFeatures.ScmHA.chooseContainerPathID(this, getClusterID() == null ? "" : getClusterID());`
4.) In `keyValueContainerUtil.getLeftovers` and `keyValueContainerUtil.moveToTmpDirectory` call `createDeleteServiceDir `instead of `checkTmpPaths`.  The `createDeleteServiceDir` will check whether the path exists in hddsvolume intstance or on disk as we want it to.

With the above changes, the code is simplified in the datanode start `VersionEndpointTask`, we no longer need the checkTmpPaths, it takes into account the volume path layout for different versions (scmid, clusterid) and it seems to pass all the unit tests.  I'll supply you with the diff.  Note the unit test failures in the current CI workflow.

xBis7 replied:
@neils-dev This approach is simpler and seems to work as expected. Thanks!

xBis7 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:
```diff
@@ -395,4 +413,179 @@ public static Path getMetadataDirectory(
+        LOG.error("Volume hasn't been formatted " +
+            "properly and has no ClusterId. " +
+            "Unable to initialize tmp and delete service directories.");
+      }
```
There are no issues. I proceeded with the changes.

xBis7 commented:
hey @DaveTeng0, if you are referring to the workflow, it's unrelated to the PR. I got a green workflow for the latest 2 commits on my fork. Check below

[cleanup](https://github.com/xBis7/ozone/actions/runs/3876726389)

[remove null check before setting tmp dir paths](https://github.com/xBis7/ozone/actions/runs/3876633980)

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:
```diff
@@ -33,24 +47,48 @@
+  private final HddsDatanodeService service =
+      HddsDatanodeService.createHddsDatanodeService(new String[] {});
+  private static final int SCM_SERVER_COUNT = 1;
+  private static final String FILE_SEPARATOR = File.separator;
```
Would be better to use Ozone separator delimiter: `OzoneConts.OZONE_URI_DELIMITER `

neils-dev commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:
```diff
@@ -59,6 +67,8 @@ private KeyValueContainerUtil() {
   private static final Logger LOG = LoggerFactory.getLogger(
       KeyValueContainerUtil.class);
 
+  private static final String FILE_SEPARATOR = File.separator;
```
Ozone system file separator can be used here: `OZONE_URI_DELIMITER`.

neils-dev commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:
```diff
@@ -395,4 +413,158 @@ public static Path getMetadataDirectory(
+          // Remove container from Rocks DB
+          BlockUtils.removeContainerFromDB(keyValueContainerData,
+              hddsVolume.getConf());
+        }
```
The `removeContainerFromDB` can throw an exception should the delete from the volume DB fail.  We should catch this here, _log the error_, and continue iterating through the list of containers to delete.  On the next call for this volume, the failed container db entry will be retried to remove from the db and subsequently delete the container should it succeed.

neils-dev commented:
In the `KeyValueHandler.importContainerData`, it too tries to cleanup the container should the import fail through container delete.  This would be a good candidate for the container _move_ to `tmp` directory and delete to ensure no artifacts are left in case the delete fails.  https://github.com/apache/ozone/blob/6389d01605053a9a5ae5510b67fb2bc6204db9d2/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueContainer.java#L560

@errose28 , is this what you had in mind for reusing the `tmp dir` and cleanup for the volume?  In this case should the container import fails, it would move the container to the `tmp `directory, try to delete from db.  In the event of failure it should be picked up by the cleanup on datanode shutdown/start and retried.

xBis7 commented:
@neils-dev I've updated the patch.

neils-dev commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/HddsVolume.java:
```diff
@@ -350,6 +368,63 @@ public void createDbStore(MutableVolumeSet dbVolumeSet) throws IOException {
+   * 
+   * @param id clusterId or scmId
+   */
+  public void checkTmpDirPaths(String id) {
```
With restructuring the location of configuring the tmp directory into the `HddsVolume`, passing the `clusterID` to the tmp dir path config is redundant.  _Remove_ all interfaces with `id `(clusterid) as an argument as the `clusterID `is a private variable in the `HddsVolume `(`StorageVolume`) instance (use `getClusterID`).

neils-dev commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/HddsVolume.java:
```diff
@@ -350,6 +368,63 @@ public void createDbStore(MutableVolumeSet dbVolumeSet) throws IOException {
+    String vol = HddsVolumeUtil.getHddsRoot(hddsRoot);
+
+    Path volPath = Paths.get(vol);
+    Path idPath = Paths.get(id);
```
The id is the <clusterid / scmid> based on the version.  It may be best to check and set the `id` from `VersionDatanodeFeatures.chooseContainerPathID` here.  Once set here, all calls to the tmp dir container delete service path can be independent of the version, having properly set the path correctly in one place.

xBis7 replied:
I've been looking at all the usage and [this](https://github.com/apache/ozone/blob/f291058fa8974fe3a2844df5d27ec444e9826742/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/StorageVolumeUtil.java#L241-L265) seems the only case where clusterID might not be available and we need to set it. 

For calling `VersionDatanodeFeatures.ScmHA.chooseContainerPathID` we need to know clusterId which might not be available in the corner case we are trying to cover here.

It makes more sense to me to overload `checkTmpDirPaths()` and have a separate `checkTmpDirPaths(String id)` where we choose the id in StorageVolumeUtil and pass it as a parameter.  

@neils-dev What do you think?

neils-dev replied:
We should be able to use `chooseContainerPathID(StorageVolume volume, String clusterID) `inside the original `checkTmpDirPaths` so one place sets the proper location for the subdir of hdds.  The `chooseContainerPathID `should resolve the proper location of the hdds subdir independent of version.  Inside `checkTmpDirPaths`, just check if id is null, if so pass an emptyString for `clusterID `to `chooseContainerPathId `and it should return the proper location of the hdds subdir (scmid or clusterid).  Let me know how it goes.

xBis7 replied:
This is handled in the new method `getIdDir()`. I'll resolve this conversation.

neils-dev commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/HddsVolume.java:
```diff
@@ -350,6 +368,63 @@ public void createDbStore(MutableVolumeSet dbVolumeSet) throws IOException {
+    return volPath.resolve(idPath).resolve(TMP_DIR);
+  }
+
+  private void createTmpDir(String id) {
```
update interface, unnecessary `id` argument. (clusterid defined in HddsVolume / StorageVolume)

neils-dev commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/HddsVolume.java:
```diff
@@ -350,6 +368,63 @@ public void createDbStore(MutableVolumeSet dbVolumeSet) throws IOException {
+    }
+  }
+
+  public void createDeleteServiceDir(String id) {
```
update interface, unnecessary `id` argument.  (clusterid defined in HddsVolume / StorageVolume)

neils-dev commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:
```diff
@@ -1236,6 +1237,29 @@ private void deleteInternal(Container container, boolean force)
+        boolean success = KeyValueContainerUtil.ContainerDeleteDirectory
+            .moveToTmpDeleteDirectory(keyValueContainerData, hddsVolume);
+
+        if (success) {
```
This block (success) can be restructured to only be used and output only if debug logging is enabled.  Something like:

```
        if (!success) {
          LOG.error("Failed to move container under " +
              hddsVolume.getDeleteServiceDirPath());
          throw new StorageContainerException("Moving container failed",
              CONTAINER_INTERNAL_ERROR);
        }
        if (LOG.isDebugEnabled()) {
          String containerPath = keyValueContainerData
              .getContainerPath().toString();
          File containerDir = new File(containerPath);

          LOG.debug("Container {} has been successfuly moved under {}",
              containerDir.getName(), hddsVolume.getDeleteServiceDirPath());
        }
```

neils-dev reviewed (commented):
Thanks @xBis7 for your updates.  I have a few comments that came up as we wrap up reviewing this PR.  Still going over some comments left by @errose28 and want to look at the tests.

neils-dev commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/helpers/KeyValueContainerUtil.java:
```diff
@@ -395,4 +413,162 @@ public static Path getMetadataDirectory(
+      }
+
+      // Initialize tmp and delete service directories
+      hddsVolume.checkTmpDirPaths(hddsVolume.getClusterID());
```
The `checkTmpDirPaths` is called again in the line below within the `getDeleteLeftovers`.  In this case it is called twice.  Where is it better placed?

xBis7 replied:
I'll remove this check here and replace it with a comment.

xBis7 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/HddsDatanodeService.java:
```diff
@@ -116,8 +117,25 @@ public HddsDatanodeService(boolean printBanner, String[] args) {
+    if (datanodeStateMachine != null) {
+      MutableVolumeSet volumeSet =
+          datanodeStateMachine.getContainer().getVolumeSet();
+      for (HddsVolume hddsVolume : StorageVolumeUtil.getHddsVolumesList(
```
@neils-dev In `ScmHAFinalizeUpgradeActionDatanode` we are iterating `volumeSet.getVolumesList()` which returns all volumes and in that case, it makes sense to check their instance. 

Here, we are wraping the `volumeSet.getVolumesList()` with `StorageVolumeUtil.getHddsVolumesList(volume list)` which filters the volumes and returns only the HddsVolumes.

https://github.com/apache/ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/StorageVolumeUtil.java#L66

neils-dev replied:
Thanks @xBis7 , that util, it seems to _cast_ each entry in the volume stream to an `HddsVolume` and returns the corresponding list.  It would be better to iterate the volume set and only use `HddsVolume`.

xBis7 replied:
I didn't realize it was casting to `HddsVolume`. You are right, in that case, it's best to iterate all volumes and check the instance. I'll fix it.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:
```diff
@@ -88,6 +125,44 @@ public void testStartup() throws IOException {
+    // in order to access them after service.stop()
+    MutableVolumeSet volumeSet = service
+        .getDatanodeStateMachine().getContainer().getVolumeSet();
+    List<HddsVolume> volumes = StorageVolumeUtil.getHddsVolumesList(
```
Similar to previous comment.  Good to check each volume in list instanceof `HddsVolume`.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:
```diff
@@ -88,6 +125,44 @@ public void testStartup() throws IOException {
+      // Create a container and move it under the tmp delete dir.
+      KeyValueContainer container = ContainerTestUtils
+          .setUpTestContainer(volume, clusterId, conf, schemaVersion);
+      assertTrue(container.getContainerFile().exists());
```
For Schema_V3, an added entry into the db and deleting entry based on container in tmp directory would be a great addition.

xBis7 replied:
@neils-dev I've added the test, but it's on `TestEndPoint`. `TestHddsDatanodeService` tests tmp dir cleanup on shutdown where db is closed, so I added the check on startup where we don't have such issue.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestContainerPersistence.java:
```diff
@@ -286,19 +288,161 @@ public void testDeleteContainer() throws Exception {
+
+    // With schema v3, we don't have a container dedicated db,
+    // so skip check the behaviors related to it.
+    assumeFalse(schemaVersion.contains(OzoneConsts.SCHEMA_V3));
```
Does this test not work with schema V3 as well.  It is testing deleting a non-empty container, which is independent of the db schema (whether db per datanode or db per volume).

xBis7 replied:
If you check the diffs, this was part of a bigger test that I refactored and it was returning in case we had SchemaV3 so I didn't check it out or made any changes. I just broke it into smaller tests.

You are right, it works regardless of Schema.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestContainerPersistence.java:
```diff
@@ -286,19 +288,161 @@ public void testDeleteContainer() throws Exception {
+  public void testDeleteContainerWithRenaming()
+      throws Exception {
+    HddsVolume hddsVolume = null;
+    if (!schemaVersion.contains(OzoneConsts.SCHEMA_V3)) {
```
Are we not handling container delete with rename for all schemas now?  If so, this test does the same thing 3 times, once for each schema without altering behavior.  The differences between < V3 and V3 was the db delete.  It doesn;t look to be involved in this test.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestContainerPersistence.java:
```diff
@@ -286,19 +288,161 @@ public void testDeleteContainer() throws Exception {
+        (KeyValueContainerData) container2.getContainerData();
+
+    if (schemaVersion.contains(OzoneConsts.SCHEMA_V3)) {
+      hddsVolume = container1Data.getVolume();
```
Is this condition needed?

xBis7 replied:
It seems to pass without it.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/TestHddsDatanodeService.java:
```diff
@@ -48,10 +48,7 @@
 import static org.apache.hadoop.hdds.HddsConfigKeys.HDDS_CONTAINER_TOKEN_ENABLED;
 import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_SECURITY_ENABLED_KEY;
 import static org.apache.hadoop.ozone.OzoneConsts.OZONE_URI_DELIMITER;
-import static org.junit.jupiter.api.Assertions.assertFalse;
```
Having the Junit 5 assert methods fully qualified is fine.  We can leave them in the file and prevent unneeded changes having to be merged.  We can leave the fully qualified imports for `assertFalse`, `assertTrue`, `assertNotNull`, `assertNull `and revert the changes in the file due to their omission.

xBis7 replied:
I know that it seems we are creating unnecessary diffs here but it's much cleaner having one import instead of five. Apart from the existing, we will add 2 more (`assertEquals`, `assertTrue`) and generate new diffs anyway.

```java
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertEquals;
```

neils-dev replied:
Let's just revert this for now.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/ContainerTestUtils.java:
```diff
@@ -207,16 +207,21 @@ public static KeyValueContainer setUpTestContainer(
         UUID.randomUUID().toString(),
         UUID.randomUUID().toString());
     keyValueContainerData.setSchemaVersion(schemaVersion);
+    keyValueContainerData.setState(ContainerProtos
```
Is this change needed in the tests?  I removed this and the `parseKVContainerData` addition to the test util without affecting the tests.  We can omit them if not needed for tests.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/ContainerTestUtils.java:
```diff
@@ -207,16 +207,21 @@ public static KeyValueContainer setUpTestContainer(
 
+    KeyValueContainerUtil
+        .parseKVContainerData(keyValueContainerData, conf);
+
```
Same as prev comment, needed in test util for tests?  If not, we can omit.

neils-dev commented on hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/impl/TestContainerPersistence.java:
```diff
@@ -267,45 +267,45 @@ public void testAddingBlockToDeletedContainer() throws Exception {
     assumeFalse(schemaVersion.contains(OzoneConsts.SCHEMA_V3));
 
-    long testContainerID1 = getTestContainerID();
+    long testContainerID = getTestContainerID();
```
Pls try to keep changes made to code under review to just address review comments.

neils-dev commented on hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/ozone/container/common/TestEndPoint.java:
```diff
@@ -178,32 +187,61 @@ public void testTmpDirCleanup() throws Exception {
 
-      Assertions.assertEquals(EndpointStateMachine.EndPointStates.REGISTER,
-          newState);
+      StorageVolumeUtil.checkVolume(hddsVolume, clusterId,
```
We should re-use the `ContainerTestUtils.createDbInstancesForTestIfNeeded` to setup test here instead of the `checkVolume`.

neils-dev commented:
@errose28 , we've raised a ticket on an issue we found while performing end to end tests with Container Delete with `V3 Schema Rocks db`,   https://issues.apache.org/jira/browse/HDDS-8173 .

On container delete processed by the KeyValueHandler.deleteInternal and handled by the `KeyValueContainerUtil.removeContainer`, the rocksdb V3 entries for the container are not cleared.  After container delete the container metadata and chunks directories are deleted however rocksdb V3 container table entries are not removed. 

Within the filed ticket for that issue a Test is included (patch to `TestContainerPersistence`) can be used to verify the behavior rocksdb V3 container delete behavior.  https://issues.apache.org/jira/secure/attachment/13056380/rocksDBContainerDelete.diff

`Schema V2` rocksdb should be unaffected as there is a rocksDB per container.  On container delete, the rocksDB is moved to the tmp directory along with the container metadata and chunks directory and deleted.

neils-dev commented:
Other than the issue raised in HDDS-8173, the end to end testing of container delete with atomic move works without issue.  Normal container delete, attempts to remove container entries from `rocksdb (schema V3)`, moves the container directories on the volume to the tmp directory and deletes.  Should an error occur either from an exception generated by `rocksDB `helper functions or from the deleting the `tmp` directory after move, the delete from `rocksdb (schema V3) `and delete from `tmp` is retried on shutdown of datanode and restart.

neils-dev reviewed (approved):
Merging.  Thanks @xBis7  for your work over the course of this PR.  Thanks for your review comments, suggestions, and offline discussions @errose28, @nandakumar131 , @kerneltime , @captainzmc , @myskov , @swagle, @adoroszlai .

adoroszlai commented:
@neils-dev please trim "extended description" when merging PR.  I think listing all individual commit messages from the PR (which Github does by default) is unnecessary, and in this case it is really overly verbose:


```
    HDDS-6449. Failed container delete can leave artifacts on disk (#3741)
    
    * cleanUpManager - tmpDir - rename container path - delete container in tmpDir
    
    * TestContainerDeleteWithSchemaV3Enabled improvement, cleanup
    
    * cleanTmpDir on datanode start and shutdown, cleanup
    
    * File.renameTo - rename before deletes if V3 enabled
    
    * datanode service start/stop cleanTmpDir()
    
    * HddsDatanodeService cleanTmpDir()  getting schema
    
    * testing cleanTmpDir on datanode start/stop
    
    * modify /tmp path based on volume
    
    * adding cluster id to /tmp path
    
    * remove flag SchemaV3Enabled - clusterId/scmId on the /tmp path
    
    * move cleanTmpDir startup hook to VersionEndpointTask
    
    * no changes
    
    * checking if schema v3 enabled
    
    * check schema v3 is enabled using configuration in CleanUpManager
    
    * initialize CleanUpManager if schemaV3 - testDeleteContainerWithSchemaV3Enabled execute conditionally
    
    * TestHddsDatanodeService - getTmpDirPath from CleanUpManager
    
    * TestKeyValueHandler testDeleteContainerWithSchemaV3Enabled
    
    * TestContainerPersistence fix
    
    * findbugs errors fixed
    
    * TestEndpoint test clean tmp dir on datanode startup
    
    * TestContainerPersistence fix - datanode startup hook moved
    
    * checkstyle fix
    
    * failing tests fixed

    * cleanup
    
    * cleanup - moving CleanUpManager under common.helpers package
    
    * check instance of containerdata - cleanup
    
    * TestContainerPersistence findbugs error fix
    
    * CleanUpManager.renameDir javadoc comment
    
    * CleanUpManager.setTmpDirPath exception message
    
    * CleanUpManager tmpDirPath made final
    
    * moving tmp dir under HddsVolume
    
    * IntegrationTest error fixed
    
    * deleted tmp files
    
    * TestDatanodeUpgradeToScmHA#testImportContainer fix
    
    * cleanup
    
    * unit tests fixed
    
    * checkstyle fixed
    
    * remove container db failure exception
    
    * db failure rethrow exception
    
    * cleanup
    
    * db failure exception
    
    * container delete operations moved under KeyValueContainerUtil
    
    * workflow errors fixed
    
    * rebasing - TestHddsDatanodeService migrating to junit5
    
    * build errors from rebasing, fixed - KeyValueContainerUtil cleanup
    
    * tests cleaned up
    
    * findbugs error fixed
    
    * null pointer exception for tmp dir during datanode startup fixed
    
    * calling ContainerUtils#getContainerFile from cleanTmpDir - tests updated
    
    * scm exception if container move failed
    
    * rebasing
    
    * createTmpDir if null, while creating delete service dir
    
    * deleteInternal more specific SCE thrown on container rename
    
    * HddsVolume using Paths lib - cleanup
    
    * remove null check before setting tmp dir paths
    
    * cleanup
    
    * logging error on removeContainerFromDB failure
    
    * cleanup
    
    * removing checkTmpDirs() - tests cleanup
    
    * delete test generated directories
    
    * tests cleanup
    
    * test db entries on tmp dir cleanup
    
    * merging errors fixed
    
    * tests cleanup
    
    * reducing test diffs
    
    ---------
```
